% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/circRNAFull.R
\name{Extract_exon}
\alias{Extract_exon}
\title{Extracting exon for the circRNAs}
\usage{
Extract_exon(bed_file, folder_name, circle_ids, output_folder)
}
\arguments{
\item{bed_file}{A data frame containing the exon coordinated obtained from reference genome annotation}

\item{folder_name}{Location of the individual alignments (obtained from function \code{\link[circRNAFull]{extract_reads_from_bam}})}

\item{circle_ids}{A data frame containing the circle id, transcript name and spanning reads (obtained from function \code{\link[circRNAFull]{extract_circle_ids}})}

\item{output_folder}{The output folder name}
}
\value{
Exons for circRNAs
}
\description{
This function extracts exons for the circRNAs from the CIGAR value of the chimeric alignment
}
\examples{
#Loading an example bed_file
data(exon_coordinate)
bed_file<-exon_coordinate

#Example of folder_name containing the individual alignments
folder_name<-system.file('extdata', package = 'circRNAFull')

#Loading an example circle_ids file
data(circle_ID)
circle_ids<-circle_ID

#creating a temporary output_folder
output_folder<-tempdir()

#Extracting exon for all circRNAs
Extract_exon(bed_file, folder_name, circle_ids, output_folder)
}
